### PR TITLE
[FLINK-30397][Connector/Pulsar] Remove Pulsar connector

### PR DESCRIPTION
**Pull request by MartijnVisser**
## What is the purpose of the change

The Pulsar connector code has been moved to https://github.com/apache/flink-connector-pulsar. 
* The `3.0` branch contains the 1.16 code, which has been officially released.
* The current `main` branch contains the code from this repo's `master` branch. 

I believe we can already remove the code from the repository, because the current code from `master` (or `main` if you look at the externalized connector repo) is a breaking change compared to the Flink 1.16 release and can only be released when Flink 1.17 has been released. Given that circumstances plus the instability of the Pulsar connector on the overall Flink CI stability would make it good to already remove the connector. 

## Brief change log

* Removed all source code, documentation and references
* Updated the Python code to refer to the externalized connector version

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java [hunk 1]
```diff
@@ -73,15 +73,6 @@ public class ConfigurationOptionLocator {
...
                         "flink-table/flink-sql-client", "org.apache.flink.table.client.config"),
-                new OptionsClassLocation(
-                        "flink-connectors/flink-connector-pulsar",
-                        "org.apache.flink.connector.pulsar.common.config"),
-                new OptionsClassLocation(
-                        "flink-connectors/flink-connector-pulsar",
-                        "org.apache.flink.connector.pulsar.source"),
-                new OptionsClassLocation(
-                        "flink-connectors/flink-connector-pulsar",
-                        "org.apache.flink.connector.pulsar.sink"),
```
Comment (zentol) [inline flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java:84]: The associated generated files should be removed. The pulsar docs will have to provide equivalent files _somehow_.
(medium-term we need to make the config docs generator available to connector repos)

File: mvnw [hunk 2]
```diff
@@ -1,316 +0,0 @@
-#!/bin/sh
-# ----------------------------------------------------------------------------
-# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (zentol) [inline mvnw:3]: This shouldnt be deleted?

File: tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/JarFileChecker.java [hunk 3]
```diff
@@ -210,11 +210,6 @@ private static int findNonBinaryFilesContainingText(
                     // dual-licensed under GPL 2 and EPL 2.0
                     // contained in sql-avro-confluent-registry
                     .filter(path -> !pathStartsWith(path, "/org/glassfish/jersey/internal"))
-                    // contained in sql-connector-pulsar
-                    .filter(
-                            path ->
-                                    !pathStartsWith(
-                                            path, "/org/apache/pulsar/shade/org/glassfish/jersey/"))
```
Comment (zentol) [inline tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/JarFileChecker.java:217]: We need to keep this for now, as otherwise CI will fail in the pulsar repo (eventually).

**Review by zentol — CHANGES_REQUESTED** [review]
you should update the production tests archunit store to remove pulsar violations.

File: mvnw (hunk 2 above)
Comment (MartijnVisser) [reply mvnw:3]: Honestly don't know why that happened, reverted

File: flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java (hunk 1 above)
Comment (MartijnVisser) [reply flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java]: I would think that right now we should 'just' remove it here and create a follow-up ticket for a permanent solution, WDYT?

**Comment by MartijnVisser** [comment]
> you should update the production tests archunit store to remove pulsar violations.

Aren't the Pulsar violations already removed, since I've completely removed the Pulsar source code? I don't see any other violations in the `flink-architecture-tests-production` violations folder?

File: flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java (hunk 1 above)
Comment (MartijnVisser) [reply flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java:84]: > The pulsar docs will have to provide equivalent files somehow.

Can we tackle that as part of a follow-up ticket?

File: docs/setup_docs.sh [hunk 4]
```diff
@@ -42,6 +42,10 @@ function integrate_connector_docs {
   theme_dir="../themes/connectors"
   mkdir -p "${theme_dir}"
   rsync -a flink-connector-${connector}/docs/content* "${theme_dir}/"
+  DIR=flink-connector-${connector}/docs/layouts;
+  if [ -e $DIR ];
+    then rsync -a flink-connector-${connector}/docs/layouts* "${theme_dir}/"
+  fi
```
Comment (XComp) [inline docs/setup_docs.sh:48]: Why is this necessary now and wasn't necessary with the other externalized connectors?

**Review by XComp — CHANGES_REQUESTED** [review]
There's still the `JarChecker` that holds some pulsar reference. I guess, that's something we could get rid off as well? I assume that we want to keep the pyflink Pulsar-related source code.

File: docs/setup_docs.sh [hunk 5]
```diff
@@ -54,6 +58,7 @@ cd tmp
 integrate_connector_docs elasticsearch v3.0.0
 integrate_connector_docs aws v3.0.0
 integrate_connector_docs cassandra v3.0.0
+integrate_connector_docs pulsar main
```
[the remark this answers is not in the document]
Comment (MartijnVisser) [reply docs/setup_docs.sh:61]: No, because the externalized Pulsar connector in `main` is only compatible with Flink 1.17 (because of FLINK-28853) so it can only be updated when Flink 1.17 is released and there has been a new Pulsar release created (v4.0.0)

**Comment by MartijnVisser** [comment]
> There's still the `JarChecker` that holds some pulsar reference. I guess, that's something we could get rid off as well? I assume that we want to keep the pyflink Pulsar-related source code.

I've tried that, but PyFlink requires that actually.

File: docs/setup_docs.sh (hunk 4 above)
Comment (MartijnVisser) [reply docs/setup_docs.sh:48]: Because the Pulsar connector has generated shortcodes for the documentation, which no other externalized connector has had (so far). See https://github.com/apache/flink-connector-pulsar/tree/main/docs/layouts/shortcodes

File: docs/setup_docs.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (XComp) [reply docs/setup_docs.sh:61]: I see, thanks for clarification.

File: docs/setup_docs.sh (hunk 4 above)
Comment (XComp) [reply docs/setup_docs.sh:48]: Shouldn't we make this then a parameter instead of hardcoding it in the function?
```
function integrate_connector_docs {
  local connector ref additional_folders
  connector=$1
  ref=$2
  additional_folders=( "${@[@]:3}" )

  git clone --single-branch --branch "${ref}" "https://github.com/apache/flink-connector-${connector}"
  theme_dir="../themes/connectors"
  mkdir -p "${theme_dir}"

  for rsync_folder_subpath in "docs/content*" "${additional_folders[@]}"; do
    local rsync_source_path="flink-connector-${connector}/${rsync_folder_subpath}"
    [ -e "${rsync_source_path}" ] || echo "Error: Invalid sub-path ${rsync_folder_subpath}! Folder does not exist." && exit 1

    rsync -a "flink-connector-${connector}/${rsync_folder_subpath}" "${theme_dir}/"
  done
}

# Integrate the connector documentation

rm -rf themes/connectors/*
rm -rf tmp
mkdir tmp
cd tmp

integrate_connector_docs elasticsearch v3.0.0
integrate_connector_docs aws v3.0.0
integrate_connector_docs cassandra v3.0.0
integrate_connector_docs pulsar main "docs/layouts*"
# [...]
```

**Review by XComp — COMMENTED** [review]
I added a proposal to one of the thread. See [my comment above](https://github.com/apache/flink/pull/21501#discussion_r1054411775). About the `JarChecker`: Could we extend the comment to the JarChecker for the pulsar connector mentioning that it's required for flink-python?

File: docs/setup_docs.sh (hunk 4 above)
Comment (MartijnVisser) [reply docs/setup_docs.sh:48]: I have no strong opinion on this, so I'll take your suggestion

**Comment by MartijnVisser** [comment]
> I added a proposal to one of the thread. See [my comment above](https://github.com/apache/flink/pull/21501#discussion_r1054411775). 

Fixed

> About the `JarChecker`: Could we extend the comment to the JarChecker for the pulsar connector mentioning that it's required for flink-python?

Added :)

File: tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/JarFileChecker.java [hunk 6]
```diff
@@ -211,6 +211,7 @@ private static int findNonBinaryFilesContainingText(
                     // contained in sql-avro-confluent-registry
                     .filter(path -> !pathStartsWith(path, "/org/glassfish/jersey/internal"))
                     // contained in sql-connector-pulsar
+                    // while the Pulsar connector is externalized, this is still needed for PuFlink
```
Comment (XComp) [inline tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/JarFileChecker.java:214]:
```suggestion
                    // while the Pulsar connector is externalized, this is still needed for PyFlink
```

File: flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java (hunk 1 above)
Comment (XComp) [reply flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java:84]: Did we create the follow-up ticket?

File: docs/setup_docs.sh [hunk 7]
```diff
@@ -36,14 +36,25 @@ echo "Created temporary file" $goModFileLocation/go.mod
...
   connector=$1
   ref=$2
+  additional_folders=( "${@:3}" )
+
   git clone --single-branch --branch ${ref} https://github.com/apache/flink-connector-${connector}
   theme_dir="../themes/connectors"
   mkdir -p "${theme_dir}"
-  rsync -a flink-connector-${connector}/docs/content* "${theme_dir}/"
+
+  for rsync_folder_subpath in "docs/content*" "${additional_folders[@]}"; do
```
Comment (XComp) [inline docs/setup_docs.sh:48]:
```suggestion
  for rsync_folder_subpath in "docs/content" "docs/content.zh" "${additional_folders[@]}"; do
```
sorry; my bad. The if statement doesn't allow the asterix. AFAIU, the asterix was just used initially to rsync both `docs/content` and `docs/content.zh`.

File: docs/setup_docs.sh [hunk 8]
```diff
@@ -54,6 +65,7 @@ cd tmp
 integrate_connector_docs elasticsearch v3.0.0
 integrate_connector_docs aws v3.0.0
 integrate_connector_docs cassandra v3.0.0
+integrate_connector_docs pulsar main "docs/layouts*"
```
Comment (XComp) [inline docs/setup_docs.sh:68]:
```suggestion
integrate_connector_docs pulsar main "docs/layouts"
```
I know that just blindly copied the `layouts*`. So, it's actually my mistake. Sorry for that. But am I right in assuming that `docs/layouts*` was just copied from `docs/content*` without a reason? Or why do we need the asterix here? It appears that there is only a `docs/layout` subfolder in [flink-connector-pulsar:./docs/](https://github.com/apache/flink-connector-pulsar/tree/main/docs).

**Review by XComp — COMMENTED** [review]
I added some comments. PTAL

File: docs/setup_docs.sh (hunk 7 above)
Comment (MartijnVisser) [reply docs/setup_docs.sh:48]: This was also sloppy on my part, should have tried it locally first 😅

File: docs/setup_docs.sh (hunk 8 above)
Comment (MartijnVisser) [reply docs/setup_docs.sh:68]: > But am I right in assuming that docs/layouts* was just copied from docs/content* without a reason? 

Correct :) Fixed

File: flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java (hunk 1 above)
Comment (MartijnVisser) [reply flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java:84]: It is now in FLINK-30486

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Review by XComp — APPROVED** [review]
LGTM now. Thanks for addressing my comments.
